Add fastpath for dot and vdot when the inputs have conj bit set to True

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#62915 Add fastpath for dot and vdot when the inputs have conj bit set to True**

As much as 45% and 20% perf improvement on CUDA and CPU respectively. 
consistent improvement in perf for all cases -- see perf numbers in comments below

Differential Revision: [D30404006](https://our.internmc.facebook.com/intern/diff/D30404006)

anjali411 committed:
Add fastpath for dot and vdot when the inputs have conj bit set to True

[ghstack-poisoned]

anjali411 commented:
Comparing the following calls: 
1. 
```
vec1 = torch.randn(m, dtype=dtype, device=device)
vec2 = torch.randn(m, dtype=dtype, device=device)
task1 = "torch.vdot(vec1.conj_physical(), vec2.conj_physical())"
task2 = "torch.vdot(vec1.conj(), vec2.conj())"
```

<details open>
<summary> Benchmark results: </summary>

[ vdot(vec1.conj_physical(), vec2.conj_physical()) input dtype:torch.complex128 device:cpu ]
           |  Using conj_physical
1 threads: ----------------------
      2    |          6.7
      8    |          6.7
      64   |          6.9
      128  |          6.9
      256  |          7.4
      512  |          7.9

Times are in microseconds (us).

[ vdot(vec1.conj(), vec2.conj()) input dtype:torch.complex128 device:cpu ]
           |  Using conjugate view
1 threads: -----------------------
      2    |          7.8
      8    |          7.9
      64   |          8.2
      128  |          8.2
      256  |          8.5
      512  |          8.7

Times are in microseconds (us).

[ vdot(vec1.conj_physical(), vec2.conj_physical()) input dtype:torch.complex128 device:cuda ]
           |  Using conj_physical
1 threads: ----------------------
      2    |          71.2
      8    |          44.3
      64   |          43.9
      128  |          45.1
      256  |          45.6
      512  |          46.2

Times are in microseconds (us).

[ vdot(vec1.conj(), vec2.conj()) input dtype:torch.complex128 device:cuda ]
           |  Using conjugate view
1 threads: -----------------------
      2    |          41.4
      8    |          42.2
      64   |          43.1
      128  |          42.9
      256  |          43.6
      512  |          43.8

Times are in microseconds (us).

</details>

2.
```
task1 = "torch.vdot(vec1, vec2.conj_physical())"
task2 = "torch.vdot(vec1, vec2.conj())"
```
<details open>
<summary> Benchmark results: </summary>

[ vdot(vec1, vec2.conj_physical()) input dtype:torch.complex128 device:cpu ]
           |  Using conj_physical
1 threads: ----------------------
      2    |          7.4
      8    |          7.5
      64   |          5.1
      128  |          7.9
      256  |          8.1
      512  |          5.4

Times are in microseconds (us).

[ vdot(vec1, vec2.conj()) input dtype:torch.complex128 device:cpu ]
           |  Using conjugate view
1 threads: -----------------------
      2    |          7.5
      8    |          4.8
      64   |          4.7
      128  |          7.6
      256  |          7.5
      512  |          8.1

Times are in microseconds (us).

[ vdot(vec1, vec2.conj_physical()) input dtype:torch.complex128 device:cuda ]
           |  Using conj_physical
1 threads: ----------------------
      2    |          56.2
      8    |          40.3
      64   |          31.1
      128  |          31.0
      256  |          30.6
      512  |          30.3

Times are in microseconds (us).

[ vdot(vec1, vec2.conj()) input dtype:torch.complex128 device:cuda ]
           |  Using conjugate view
1 threads: -----------------------
      2    |          21.7
      8    |          21.9
      64   |          21.7
      128  |          21.2
      256  |          22.1
      512  |          21.3

Times are in microseconds (us).

</details>

3. 
```
task1 = "torch.vdot(vec1.conj_physical(), vec2)"
task2 = "torch.vdot(vec1.conj(), vec2)"
```
<details open>
<summary> Benchmark results: </summary>
[ vdot(vec1.conj_physical(), vec2) input dtype:torch.complex128 device:cpu ]
           |  Using conj_physical
1 threads: ----------------------
      2    |          4.7
      8    |          4.6
      64   |          4.7
      128  |          4.8
      256  |          8.2
      512  |          5.4

Times are in microseconds (us).

[ vdot(vec1.conj(), vec2) input dtype:torch.complex128 device:cpu ]
           |  Using conjugate view
1 threads: -----------------------
      2    |          5.5
      8    |          4.7
      64   |          4.6
      128  |          7.7
      256  |          7.8
      512  |          4.9

Times are in microseconds (us).

[ vdot(vec1.conj_physical(), vec2) input dtype:torch.complex128 device:cuda ]
           |  Using conj_physical
1 threads: ----------------------
      2    |          54.0
      8    |          31.2
      64   |          30.5
      128  |          30.1
      256  |          30.7
      512  |          31.7

Times are in microseconds (us).

[ vdot(vec1.conj(), vec2) input dtype:torch.complex128 device:cuda ]
           |  Using conjugate view
1 threads: -----------------------
      2    |          22.3
      8    |          22.4
      64   |          22.0
      128  |          21.1
      256  |          22.3
      512  |          22.8

Times are in microseconds (us).
</details>

anjali411 commented:
Comparing the following calls:
1.
```
vec1 = torch.randn(m, dtype=dtype, device=device)
vec2 = torch.randn(m, dtype=dtype, device=device)
task1 = "torch.dot(vec1.conj_physical(), vec2.conj_physical())"
task2 = "torch.dot(vec1.conj(), vec2.conj())"
```
<details open>
<summary> Benchmark results: </summary>

[ dot(vec1.conj_physical(), vec2.conj_physical()) input dtype:torch.complex128 device:cpu ]
           |  Using conj_physical
1 threads: ----------------------
      2    |          7.8
      8    |          8.0
      64   |          8.0
      128  |          8.1
      256  |          8.4
      512  |          9.2

Times are in microseconds (us).

[ dot(vec1.conj(), vec2.conj()) input dtype:torch.complex128 device:cpu ]
           |  Using conjugate view
1 threads: -----------------------
      2    |          8.6
      8    |          8.5
      64   |          8.4
      128  |          8.8
      256  |          9.0
      512  |          9.2

Times are in microseconds (us).

[ dot(vec1.conj_physical(), vec2.conj_physical()) input dtype:torch.complex128 device:cuda ]
           |  Using conj_physical
1 threads: ----------------------
      2    |          63.8
      8    |          43.6
      64   |          42.3
      128  |          44.8
      256  |          41.6
      512  |          41.7

Times are in microseconds (us).

[ dot(vec1.conj(), vec2.conj()) input dtype:torch.complex128 device:cuda ]
           |  Using conjugate view
1 threads: -----------------------
      2    |          39.0
      8    |          42.3
      64   |          40.4
      128  |          40.6
      256  |          39.5
      512  |          40.6

Times are in microseconds (us).

</details>

2.
```
task1 = "torch.dot(vec1, vec2.conj_physical())"
task2 = "torch.dot(vec1, vec2.conj())"
```
<details open>
<summary> Benchmark results: </summary>

[ dot(vec1.conj_physical(), vec2.conj_physical()) input dtype:torch.complex128 device:cpu ]
           |  Using conj_physical
1 threads: ----------------------
      2    |          6.7
      8    |          6.6
      64   |          6.8
      128  |          7.0
      256  |          7.2
      512  |          7.8

Times are in microseconds (us).

[ dot(vec1.conj(), vec2.conj()) input dtype:torch.complex128 device:cpu ]
           |  Using conjugate view
1 threads: -----------------------
      2    |          7.2
      8    |          7.3
      64   |          7.5
      128  |          7.5
      256  |          7.7
      512  |          8.5

Times are in microseconds (us).

[ dot(vec1.conj_physical(), vec2.conj_physical()) input dtype:torch.complex128 device:cuda ]
           |  Using conj_physical
1 threads: ----------------------
      2    |          67.2
      8    |          60.4
      64   |          42.3
      128  |          43.5
      256  |          43.4
      512  |          43.8

Times are in microseconds (us).

[ dot(vec1.conj(), vec2.conj()) input dtype:torch.complex128 device:cuda ]
           |  Using conjugate view
1 threads: -----------------------
      2    |          42.1
      8    |          41.5
      64   |          42.0
      128  |          41.9
      256  |          43.6
      512  |          42.2

Times are in microseconds (us).

</details>

3.
```
task1 = "torch.dot(vec1.conj_physical(), vec2)"
task2 = "torch.dot(vec1.conj(), vec2)"
```
<details open>
<summary> Benchmark results: </summary>
[ dot(vec1.conj_physical(), vec2) input dtype:torch.complex128 device:cpu ]
           |  Using conj_physical
1 threads: ----------------------
      2    |          4.5
      8    |          4.6
      64   |          4.7
      128  |          4.6
      256  |          4.8
      512  |          5.3

Times are in microseconds (us).

[ dot(vec1.conj(), vec2) input dtype:torch.complex128 device:cpu ]
           |  Using conjugate view
1 threads: -----------------------
      2    |          4.1
      8    |          4.1
      64   |          4.2
      128  |          4.1
      256  |          4.2
      512  |          4.4

Times are in microseconds (us).

[ dot(vec1.conj_physical(), vec2) input dtype:torch.complex128 device:cuda ]
           |  Using conj_physical
1 threads: ----------------------
      2    |          54.9
      8    |          31.5
      64   |          38.6
      128  |          33.4
      256  |          34.0
      512  |          33.1

Times are in microseconds (us).

[ dot(vec1.conj(), vec2) input dtype:torch.complex128 device:cuda ]
           |  Using conjugate view
1 threads: -----------------------
      2    |          24.0
      8    |          23.5
      64   |          25.5
      128  |          26.0
      256  |          24.9
      512  |          24.5

Times are in microseconds (us).
</details>

anjali411 committed:
Update on "Add fastpath for dot and vdot when the inputs have conj bit set to True"




[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/Blas.cpp:
```diff
@@ -152,6 +164,16 @@ Tensor vdot(const Tensor &self, const Tensor &other){
 
+  if (self.is_conj()) {
+    if (other.is_conj()) {
+      return at::vdot(self.conj(), other.clone());
```
This doesn't look right, you are still calling `vdot`, but you are conjugating the first arg. `vdot(a,b)` cannot be equal to `vdot(a_conjugated, b)`? (`.clone` on `other` shouldn't matter, it resolves conj bit, but it doesn't change logical interpretation of data in `other`, whereas `conj` on `self` changes logical interpretation of data). 
Do we have tests that would check cross product of dot/vdot with conj bit set or not set on either argument?

ezyang replied:
vdot(conj(A), conj(B)) = sum_i(a_i * conj(b_i)). So in fact you can compute this as vdot(B, A)

anjali411 replied:
yeah totally. I have local changes to update it to that. just haven't pushed them yet

ezyang commented on aten/src/ATen/native/Blas.cpp:
```diff
@@ -137,6 +137,18 @@ Tensor dot(const Tensor &self, const Tensor &other){
+  if (self.is_complex()) {
+    if (self.is_conj()) {
+      if (other.is_conj()) {
+        return at::vdot(self.conj(), other.clone());
```
since you're redispatching to calls with error checking, it would be better to do the rerouting *before* you do error checking

ezyang replied:
Instead of cloning here, you should simply conjugate the result. dot(conj(A), conj(B)) = conj(dot(A, B)) (via distributivity of complex conjugation). As the output tensor is typically much smaller than the inputs this will be faster.

anjali411 replied:
makes sense! 
what do you think about returning `tensor.conj()` vs `tensor.conj_physical()`. I worry that it might be a little unexpected for the users to get a tensor with conj bit set to `True` or `False` depending on whether one or more input tensors had their conj bit set to `True`. But it'll be faster and we did say that we want to minimize materializing the conjugation as much as possible and benefit from the conj bit optimizations wherever we can.

ezyang replied:
@Lezcano may have some thoughts here, but in this case conj physical seems pretty acceptable, tbh

Lezcano replied:
fwiw, given that the observable behaviour of tensors does not depend on the conjugate bit, I lean towards returning tensors with the conj bit on. The more operations we have that take advantage of this, the less copies we will make, which is quite nice.

ngimel replied:
For `dot`/`vdot` tbg like @ezyang I think it doesn't matter much whether we are returning an output with conjugate bit set or not, it's a scalar tensor, but for other ops returning tensors with conjugate bit set seems right (in the bright future, tensors with conjugate bit set will be first class citizens and won't need to be materialized, it's similar to how we are *not* returning contiguous tensors always, and instead propagate input strides).

ezyang replied:
One reason to advocate for returning the conjugate bit set is because dot/vdot could be batched, and then in that case there is a nontrivial perf difference. Well, maybe consistency is the important thing.

ngimel replied:
But if `dot`/`vdot` are batched they'll go through gemm/bmm, and the output will have conj bit set as returned by those ops? For consistency probably we have to have the same behavior, that's right.

ezyang commented:
Needs tests

ezyang reviewed (changes requested):
tests

anjali411 committed:
Update on "Add fastpath for dot and vdot when the inputs have conj bit set to True"




[ghstack-poisoned]

anjali411 committed:
Update on "Add fastpath for dot and vdot when the inputs have conj bit set to True"




[ghstack-poisoned]

anjali411 committed:
Update on "Add fastpath for dot and vdot when the inputs have conj bit set to True"




[ghstack-poisoned]

Lezcano commented on aten/src/ATen/native/Blas.cpp:
```diff
@@ -133,8 +133,19 @@ inline void dot_check(const Tensor& self, const Tensor& other) {
+  if (self.is_complex()) {
+    if (self.is_conj()) {
+      if (other.is_conj()) {
+        return (at::dot(self.conj(), other.conj())).conj();
```
Since `self.is_conj()` here,  would `self.conj()` incur in a copy? Would it not be better to use `_set_conj(bool)` here to avoid this copy? Same for all the other `conj()`. If this does not make a copy, feel free to mark as resolved.

anjali411 replied:
that would modify the input tensor

Lezcano replied:
Oh, I thought that it was not possible to do `self.conj()` without copying, but it's just that a new tensor is created, but no memory is allocated! Now I understand, thanks.

Lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1359,14 +1359,24 @@ def sample_inputs_bmm(self, device, dtype, requires_grad, **kwargs):
+        )
+    ))
+    if dtype.is_complex:
+        # dot/vdot for (conj(input), conj(arg_tensor)) and (conj(input), arg_tensor)
+        # is tested in test_conj_view (which tests operations with only conjugated input tensor
+        # -- not conjugated arg tensors)
```
Is the forward behaviour tested here?

ngimel replied:
`dot` doesn't have `ref` function, so its correctness is not tested, it seems.

anjali411 replied:
yeah all the three cases  `(self.conj(), other.conj())`, `(self.conj(), other.conj())`, `(self, other.conj())` for both forward and backward  will be tested in `test_conj_view`

ngimel replied:
I think going forward it would be good to modify `test_conj_view` so it conjugates the second argument also (at least for binary functions), instead of pushing this complexity to sample inputs. Then, as more operations can take advantage of the fast path, that functionality would be tested  automatically without the need to update sample inputs.

anjali411 replied:
fair point!

Lezcano commented on aten/src/ATen/native/Blas.cpp:
```diff
@@ -133,8 +133,19 @@ inline void dot_check(const Tensor& self, const Tensor& other) {
+         return at::vdot(self.conj(), other);
+       }
+    } else if (other.is_conj()) {
+      return at::vdot(other.conj(), self);
```
In all the cases, I believe it's better to call `at::native::vdot`, as it avoid a redispatch. Same for `at::native::dot`.

anjali411 replied:
yeah good catch!

lezcano reviewed (commented):
The PR looks very good. Left a few comments regarding testing / some possible optimisations.

anjali411 committed:
Update on "Add fastpath for dot and vdot when the inputs have conj bit set to True"




[ghstack-poisoned]

anjali411 committed:
Update on "Add fastpath for dot and vdot when the inputs have conj bit set to True"


As much as 45% and 20% perf improvement on CUDA and CPU respectively. 
consistent improvement in perf for all cases -- see perf numbers in comments below

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30404006).

ngimel reviewed (approved):
This LGTM! @Lezcano shout out if there is something you need addressed.

lezcano reviewed (approved):
LGTM!
